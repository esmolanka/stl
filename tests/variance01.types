
type Foo (f : -Type -> Type) -a = f a

type Foo1 -a (f : -Type -> Type) = Foo f a

type Arr -a = a -> Unit

#check {} <: {foo : Unit}

#check {foo : Unit} <: {}

#check
  Foo1 {}
  <:
  Foo1 {foo : Unit}

#check
  Foo1 {foo : Unit}
  <:
  Foo1 {}

#check
  forall (f : -Type -> Type). Foo f {}
  <:
  forall (f : -Type -> Type). Foo f {foo : Unit}

#check
  forall (f : -Type -> Type). Foo f {foo : Unit}
  <:
  forall (f : -Type -> Type). Foo f {}
