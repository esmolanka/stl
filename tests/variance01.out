tests/variance01.types:8:1-8:26:
  Subtype:
    ∃ (ρ : Row). {, … ρ}
  Supertype:
    ∃ (ρ : Row). {foo : Unit, … ρ}
  Result:
    FAIL
    Cannot unify types.
        ▫
      Is not subsumed by:
        ▪
tests/variance01.types:10:1-10:26:
  Subtype:
    ∃ (ρ : Row). {foo : Unit, … ρ}
  Supertype:
    ∃ (ρ : Row). {, … ρ}
  Result:
    OK
tests/variance01.types:12:1-15:20:
  Subtype:
    λ (f : -Type → Type). f (∃ (ρ : Row). {, … ρ})
  Supertype:
    λ (f : -Type → Type). f (∃ (ρ : Row). {foo : Unit, … ρ})
  Result:
    OK
tests/variance01.types:17:1-20:10:
  Subtype:
    λ (f : -Type → Type). f (∃ (ρ : Row). {foo : Unit, … ρ})
  Supertype:
    λ (f : -Type → Type). f (∃ (ρ : Row). {, … ρ})
  Result:
    FAIL
    Cannot unify types.
        ▫
      Is not subsumed by:
        ▪
tests/variance01.types:22:1-25:49:
  Subtype:
    ∀ (f : -Type → Type). f (∃ (ρ : Row). {, … ρ})
  Supertype:
    ∀ (f : -Type → Type). f (∃ (ρ : Row). {foo : Unit, … ρ})
  Result:
    OK
tests/variance01.types:27:1-30:39:
  Subtype:
    ∀ (f : -Type → Type). f (∃ (ρ : Row). {foo : Unit, … ρ})
  Supertype:
    ∀ (f : -Type → Type). f (∃ (ρ : Row). {, … ρ})
  Result:
    FAIL
    Cannot unify types.
        ▫
      Is not subsumed by:
        ▪